### ISSUE TITLE
PriceRounders as plugins

### ISSUE THREAD

**Issue by flathat** [CORE-POS/IS4C#1008]
* Version of CORE? 2.7.0

* Issue with Office, Lane, or both? Office

* Is this [mostly] a bug report, feature request, or question? Feature request.

I'm working on handling store-specific `PriceRounder`s as plugins, extending `\COREPOS\Fannie\API\item\PriceRounder` as an alternative to having them in the base class.

In order to decide which `PriceRounder` to use my original thought, working in `RecalculateVendorSRPs.php`, was to get a list of all of them, including the base class, which contains the default `rounder()` with:
`$price_rounders = FannieAPI::listModules('\COREPOS\Fannie\API\item\PriceRounder',True);`
and then use the first one in the array, because plugins are searched first. But I find that that only finds the plugin, and that `$price_rounders = FannieAPI::listModules('PriceRounder',True);` with the addition to `FannieAPI::searchDirectories()` of:
```PHP
            case 'PriceRounder':
                $directories[] = dirname(__FILE__).'/item/';
                break;
```
is needed to find the base class, and that if the base class is to be used, the `namespace` has to be prefixed to the name returned by `listModules()` in order for `new $prName()` to find it.

The question is a bit academic since the location of the base class is known and the literal 
`$rounder = new \COREPOS\Fannie\API\item\PriceRounder();`
can be used if there are no plugins, but I'm curious if there is a way a single call to `listModules($base,True)` could return all and `new` could find the name `listModules()` returns.

I tried putting the default `PriceRounder` in `$FANNIE/item/modules` , as a class extending the base,    
`class DefaultPriceRounder extends \COREPOS\Fannie\API\item\PriceRounder`  with the addition to `FannieAPI::searchDirectories()` of:
```PHP
            case 'COREPOS\Fannie\API\item\PriceRounder':
                $directories[] = dirname(__FILE__).'/../item/modules/';
                break;
```
and then not asking `listModules('\COREPOS\Fannie\API\item\PriceRounder',False)` to find the base. That found both the plugin and `DefaultPriceRounder` and `new $prName` found it.  Would it make sense for `CORE-POS` to actually work this way, possibly with `DefaultPriceRounder` in a different directory?

Following the model of `Accounting` and its extension `StandardAccouting` I tried putting a    
```PHP
namespace COREPOS\Fannie\API\item;
class Default2PriceRounder extends PriceRounder
```
in `classlib2.0/item/` but `listModules('PriceRounder',True) didn't find it. I'm curious why.

**Comment by gohanman** [comment]
I think you were literally a single character away from solving this yourself (although FannieAPI should handle this better).

```php
case 'PriceRounder':
                $directories[] = dirname(__FILE__).'/item/';
                break;
```
This returns something strictly because it finds a file named `PriceRounder.php` in the directory. There's probably a good argument that this special case shouldn't even exist any more. Blindly assuming a given file contains a particular class is less reliable than including that file and verifying which class(es) exist. I'm going to leave it alone though for now in case anything is relying on this behavior.

```php
case 'COREPOS\Fannie\API\item\PriceRounder':
                $directories[] = dirname(__FILE__).'/../item/modules/';
                break;
```
This one is correct and will work for `listModules('COREPOS\Fannie\API\item\PriceRounder',True)` but not `listModules('\COREPOS\Fannie\API\item\PriceRounder',True)` since in the latter case the strings don't match. That said, `listModules` should work with or without the leading backslash since both are valid ways to reference the class in PHP.

**Comment by flathat** [comment]
I assumed "solving this" meant: how to return both the plugin and base class in a single `listModules()` call.

If I understood this correctly:
> This one is correct and will work for listModules('COREPOS\Fannie\API\item\PriceRounder',True)

I expect it to return both plugins and the base class, however:
`$price_rounders = FannieAPI::listModules('COREPOS\Fannie\API\item\PriceRounder',True);`
did not (did find the plugin and the one in `$FANNIE/item/modules`).

I added the [leading-backslash removal](https://github.com/CORE-POS/IS4C/blob/f17c399ff5247b93d452492f08d01df4330456c8/fannie/classlib2.0/FannieAPI.php#L351) to `listModules()` but this:
`$price_rounders = FannieAPI::listModules('\COREPOS\Fannie\API\item\PriceRounder',True);`
also did not find the base class (but still found the others). 

Should they find both?

**Comment by gohanman** [comment]
With the fix, 11610f5, yes. For me either way of calling `listModules` finds both the base class and a subclass located in a plugin.

**Comment by gohanman** [comment]
As far as I can tell there's no upstream issue here anymore. Price Rounder can be picked up from plugins via `listModules`.
